pyomo/gdp/plugins/bigm.py

```diff
@@ -706,3 +818,57 @@ def _estimate_M(self, expr, name):
                 M = (expr_lb, expr_ub)
 
         return tuple(M)
+
+    # These are all functions to retrieve transformed components from original
+    # ones and vice versa.
+    def get_src_disjunct(self, transBlock):
+        if not hasattr(transBlock, '_srcDisjunct') or \
+           not type(transBlock._srcDisjunct) is weakref_ref:
+            raise GDP_Error("Block %s doesn't appear to be a transformation "
+                            "block for a disjunct. No source disjunct found." 
+                            % transBlock.name)
+        return transBlock._srcDisjunct()
+
+    def get_src_constraint(self, transformedConstraint):
+        transBlock = transformedConstraint.parent_block()
+        # This should be our block, so if it's not, the user messed up and gave
+        # us the wrong thing. If they happen to also have a _constraintMap then
+        # the world is really against us.
+        if not hasattr(transBlock, "_constraintMap"):
+            raise GDP_Error("Constraint %s is not a transformed constraint" 
+                            % transformedConstraint.name)
+        return transBlock._constraintMap['srcConstraints'][transformedConstraint]
+
+    def get_transformed_constraint(self, srcConstraint):
+        # We are going to have to traverse up until we find the disjunct this
+        # constraint lives on
+        parent = srcConstraint.parent_block()
+        # [ESJ 08/06/2019] I actually don't know how to do this prettily...
+        while not type(parent) in (_DisjunctData, SimpleDisjunct):
+            parent = parent.parent_block()
+            if parent is None:
+                raise GDP_Error(
+                    "Constraint %s is not on a disjunct and so was not "
+                    "transformed" % srcConstraint.name)
+        transBlock = parent._transformation_block
+        if transBlock is None:
+            raise GDP_Error("Constraint %s is on a disjunct which has not been "
+                            "transformed" % srcConstraint.name)
+        # if it's not None, it's the weakref we wanted.
+        transBlock = transBlock()
+        if hasattr(transBlock, "_constraintMap") and transBlock._constraintMap[
+                'transformedConstraints'].get(srcConstraint):
+            return transBlock._constraintMap['transformedConstraints'][
+                srcConstraint]
+        raise GDP_Error("Constraint %s has not been transformed." 
+                        % srcConstraint.name)
+
+    def get_src_disjunction(self, xor_constraint):
+        m = xor_constraint.model()
+        for disjunction in m.component_data_objects(Disjunction):
+            if disjunction._algebraic_constraint:
+                if disjunction._algebraic_constraint() is xor_constraint:
+                    return disjunction
+        raise GDP_Error("It appears that %s is not an XOR or OR constraint "
+                        "resulting from transforming a Disjunction."
+                        % xor_constraint.name)
```

Comment:
We have to do a linear scan of the model to get the source disjunction?

Reply:
Currently yes, because I didn't store a reverse map during the transformation. If we need this to be efficient, then we can just keep a map from xor constraints to their disjunctions on the transformation block. I didn't bother because I wasn't convinced a lot of people were going to be asking xor constraints where they came from. But if we ever are... Then yes, this should be changed.

Reply:
You're right. It's unlikely to be used. Maybe a quick note to that effect, so that people templating from this code understand why it was done that way?